libsolidity/ast/Types.cpp

```diff
@@ -1232,14 +1232,22 @@ pair<RationalNumberType const*, RationalNumberType const*> RationalNumberType::m
 
 	rational exponent = 0;
 	rational unsignedMantissa = abs(m_value);
+
+	if (unsignedMantissa > maxMantissa)
+		return {nullptr, TypeProvider::rationalNumber(-exponent)};
+
 	while (unsignedMantissa.denominator() != 1)
 	{
 		unsignedMantissa *= 10;
 		--exponent;
 
-		// FIXME: What happens when exponent in scientific notation is max uint?
-		if (unsignedMantissa > maxMantissa || exponent > maxUint)
+		if (exponent > maxUint && unsignedMantissa > maxMantissa)
 			return {nullptr, nullptr};
+		// FIXME: What happens when exponent in scientific notation is max uint?
+		if (exponent > maxUint)
+			return {TypeProvider::rationalNumber(unsignedMantissa), nullptr};
+		if (unsignedMantissa > maxMantissa)
+			return {nullptr, TypeProvider::rationalNumber(-exponent)};
```

Comment:
But, in this way it will always return at the first loop iteration, no?

Reply:
Ah. Right. Forgot about the `else` case :)

You could add an `if` to return only if both pair elements are non-empty.

Reply:
Sorry, could you explain this? I don't understand what you mean here.

Reply:
I meant that you would instead of returning as I suggested, you'd assign the pair to a variable, then check if both pair elements are null and return only if they're not.

But, well, maybe no point. When I commented that, I thought the `?:` version was more readable, but now going back to this, yours seems fine. Let's keep it. It's just a tiny detail anyway.